rewrites for file downloads and json output

wvanderp opened the pull request:
I am an enjoyer of bar but I was missing some critical info about the game like a comparison of the build power of various units. So I was looking into a way to parse the Lua files when I came across this project.

While I could use it for my project I saw some room for improvement. This led to a large rewrite of part of the project.

TLDR: I've made some big changes in the way the files are downloaded, the way all the different parts are glued together and I added a JSON output.

# What did not change
The command line and CSV output functions exactly the same.
I added comments to the precalculation but did not change them.
The way the Lua files are called are also mostly the same

# Some controversial changes
I applied pep-8 to the repo which means that the indentation is now with 4 spaces which means a lot of review work. Im sorry for this but I do think is better.

# GitHub
I was unhappy with the way the files are downloaded. It took a long time and when the project is run multiple times GitHub would time out.

I fixed this by using the Git command line and its sparse checkout. This is a feature of the Git client where it only downloads the necessary files. 

The big advantage is that there is no timeout. Some downsides are that you need the Git client installed (added a check for it) and it uses more disk space.

# How things are glued 

I changed the way that functions flow together

Right now the project drills down all the parts of the program before coming back up.
```
┌──────────┐
│          │
│ main     │◄────┐
│          │     │
└──────────┘     │
     ▼           │
┌──────────┐     │
│          │     │
│ download │     │
│          │     │
└──────────┘     │
     ▼           │
┌──────────┐     │
│          │     │
│ parse    │     │
│          │     │
└──────────┘     │
     ▼           │
┌──────────┐     │
│          │     │
│ output   ├─────┘
│          │
└──────────┘

```

While I made it return to main. This has a big advantage: you can use the output of functions in multiple places
For example, I used the same dictionary of units for both the JSON as well as the CSV output
```
┌────────────────────────────────────┐
│                                    │
│ main                               │
│                                    │
└────────────────────────────────────┘
  ▼     ▲      ▼      ▲      ▼    ▲
┌──────────┐   ┌─────────┐  ┌────────┐
│          │   │         │  │        │
│ download │   │ parse   │  │ output │
│          │   │         │  │        │
└──────────┘   └─────────┘  └────────┘

```
# JSON output 

The program now writes both the unprocessed units as well as the processed units to a JSON file in addition to the CSV files
I made an output directory where all the files are written to.


# Conclution 
I hope you can accept my pull request. I am available to make any changes

Teifion commented:
Given the size of it I'm confident I can't accept it tonight but this is fantastic and sounds like it implements a bunch of the things I was looking to implement at some stage. I very much hope to be either accepting this or providing feedback on it as soon as I can. Thank you for taking the time to do this!

wvanderp commented:
take your time I made it for my own use which I can now do.

Teifion commented on bar_units.py:
```diff
@@ -52,25 +54,64 @@
+    if int(major) < 2 or int(minor) < 25:
+        # you can try it with an older version git but there are no guarantees
+        raise Exception(
+            "you need a newer version of git to run this program at least 2.25.0. see code for a work around")
```
I'd like to see a better explanation of how to fix the issue in the message, especially as it doesn't say where in the code to look.

Teifion commented:
Overall I really like it except for one big thing. It can't do CSV any more. The idea was I'd be able to run it and export to CSV so I can just load it into a spreadsheet. If you can enable an option to choose the output type I'd be happy to approve the PR :)

wvanderp commented:
I added a link to the website of git. I'm not that great at writing copy so if you expected something else then I need some more guidance.

I updated the README.md to explain more about the project. this includes the new output.

the script now outputs three files two JSON and one CSV into the `generated` folder. It still takes the same parameters so ```python.exe .\bar_units.py site``` should generate the CSV file

Teifion commented:
The new exception is exactly what I was after, just something to give the person an idea of what has gone wrong. I'll do some tests and review it a little more but at this stage I expect to be accepting the PR :)
